ksql-common/src/test/java/io/confluent/ksql/schema/ksql/SqlBaseTypeTest.java

```diff
@@ -19,60 +19,97 @@
+  private static final Set<SqlBaseType> NUMBER_TYPES = ImmutableSet.of(
+      SqlBaseType.INTEGER,
+      SqlBaseType.BIGINT,
+      SqlBaseType.DECIMAL,
+      SqlBaseType.DOUBLE
+  );
 
   @Test
   public void shouldNotBeNumber() {
-    nonNumberTypes().forEach(sqlType ->
-        assertThat(sqlType + " should not be number", sqlType.isNumber(), is(false)));
+    nonNumberTypes().forEach(sqlType -> assertThat(
+        sqlType + " should not be number",
+        sqlType.isNumber(),
+        is(false)
+    ));
   }
 
   @Test
   public void shouldBeNumber() {
-    numberTypes().forEach(sqlType ->
-        assertThat(sqlType + " should be number", sqlType.isNumber(), is(true)));
+    numberTypes().forEach(sqlType -> assertThat(
+        sqlType + " should be number",
+        sqlType.isNumber(),
+        is(true)
+    ));
   }
 
   @Test
   public void shouldNotUpCastIfNotNumber() {
-    nonNumberTypes().forEach(sqlType ->
-        assertThat(sqlType + " should not upcast", sqlType.canUpCast(sqlType), is(false)));
+    nonNumberTypes().forEach(sqlType -> assertThat(
+        sqlType + " should not upcast",
+        sqlType.canUpCast(SqlBaseType.DOUBLE),
+        is(false))
+    );
   }
 
   @Test
-  public void shouldUpCastToSelfIfNumber() {
-    numberTypes().forEach(sqlType ->
+  public void shouldUpCastIfNumber() {
+    numberTypes().forEach(sqlType -> assertThat(
+        sqlType + " should upcast",
+        sqlType.canUpCast(SqlBaseType.DOUBLE),
+        is(true))
+    );
+  }
+
+  @Test
+  public void shouldUpCastToSelf() {
+    allTypes().forEach(sqlType ->
         assertThat(sqlType + " should upcast to self", sqlType.canUpCast(sqlType), is(true)));
   }
 
   @Test
   public void shouldUpCastInt() {
     assertThat(SqlBaseType.INTEGER.canUpCast(SqlBaseType.BIGINT), is(true));
+    assertThat(SqlBaseType.INTEGER.canUpCast(SqlBaseType.DECIMAL), is(true));
     assertThat(SqlBaseType.INTEGER.canUpCast(SqlBaseType.DOUBLE), is(true));
   }
 
   @Test
   public void shouldUpCastBigInt() {
+    assertThat(SqlBaseType.BIGINT.canUpCast(SqlBaseType.DECIMAL), is(true));
     assertThat(SqlBaseType.BIGINT.canUpCast(SqlBaseType.DOUBLE), is(true));
   }
 
+  @Test
+  public void shouldUpCastDecimal() {
+    assertThat(SqlBaseType.DECIMAL.canUpCast(SqlBaseType.DOUBLE), is(true));
+  }
+
   @Test
   public void shouldNotDownCastBigInt() {
     assertThat(SqlBaseType.BIGINT.canUpCast(SqlBaseType.INTEGER), is(false));
   }
 
+  @Test
+  public void shouldNotDownCastDecimal() {
+    assertThat(SqlBaseType.DECIMAL.canUpCast(SqlBaseType.INTEGER), is(false));
+    assertThat(SqlBaseType.DECIMAL.canUpCast(SqlBaseType.BIGINT), is(false));
+  }
+
   @Test
   public void shouldNotDownCastDouble() {
     assertThat(SqlBaseType.DOUBLE.canUpCast(SqlBaseType.INTEGER), is(false));
     assertThat(SqlBaseType.DOUBLE.canUpCast(SqlBaseType.BIGINT), is(false));
+    assertThat(SqlBaseType.DOUBLE.canUpCast(SqlBaseType.DECIMAL), is(false));
```

Comment:
> We should perhaps allow casting of double to decimal in the case the double is smaller than a certain value as it can be exactly represented as a decimal.

Not sure I follow: Casting of double to decimal is supported. Do you mean implicit down casting?  I don't think we can support that as the casting happens when calculating the output type of a mathematical binary operation on two _types_, not _values_.  So we can't take the _value_ into account and must always have a deterministic result _type_ based purely on the input _types_.

Again, see my comment below your review.

> BTW, imho I don't like the term "upcasting", why not just "casting"?

Not a term I introduced, this was from @agavra. Though I'm assuming it's to differentiate from down-casting, which is different.  In this context, the term up-casting is used to mean the ability for KSQL to _implicitly_ convert one number type to another type, e.g. `INT` to `BIGINT`.

Might be better if the call was called `canImplicitlyUpCast`?

Reply:
> 
> > BTW, imho I don't like the term "upcasting", why not just "casting"?
> 
> Not a term I introduced, this was from @agavra. Though I'm assuming it's to differentiate from down-casting, which is different. In this context, the term up-casting is used to mean the ability for KSQL to _implicitly_ convert one number type to another type, e.g. `INT` to `BIGINT`.
> 
> Might be better if the call was called `canImplicitlyUpCast`?

Ok, so I think "upcasting" just means "implicitly cast". (I still don't like the term 'upcasting')

And we should only do that where we can guaranteed it's safe to do so. However implicitly casting from DECIMAL to DOUBLE is not always safe. I think we currently assume it's safe because a double can hold bigger values than a decimal, but for some conversions a loss of precision might occur. Hence my previous comment about logging a warning?

Reply:
Another way of dealing with this would be to not allow DECIMAL->DOUBLE implicit conversions, and require the user to do an explicit cast. That's another way of telling the user "hey this isn't safe, but if you really want to do it, go ahead and cast"

Reply:
I think you're right about not using upcast/downcast - this was probably a terminology error on my end. Standard terminology should be `widening` and `narrowing` https://docs.oracle.com/javase/specs/jls/se7/html/jls-5.html#jls-5.1.2

Reply:
I think calling it `canImplicitlyCastTo` would be better as we're talking about _implicit_ casting, not casting in general.

Reply:
See my comment below, I think `canWiden` is probably the technically correct term.